libmachine/provision/redhat.go

```diff
@@ -0,0 +1,279 @@
+package provision
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"github.com/docker/machine/drivers"
+	"github.com/docker/machine/libmachine/auth"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/provision/pkgaction"
+	"github.com/docker/machine/libmachine/swarm"
+	"github.com/docker/machine/log"
+	"github.com/docker/machine/ssh"
+	"github.com/docker/machine/utils"
+)
+
+func init() {
+	Register("RedHat", &RegisteredProvisioner{
+		New: NewRedHatProvisioner,
+	})
+}
+
+func NewRedHatProvisioner(d drivers.Driver) Provisioner {
+	return &RedHatProvisioner{
+		packages: []string{
+			"curl",
+		},
+		Driver: d,
+	}
+}
+
+type RedHatProvisioner struct {
+	packages      []string
+	OsReleaseInfo *OsRelease
+	Driver        drivers.Driver
+	AuthOptions   auth.AuthOptions
+	EngineOptions engine.EngineOptions
+	SwarmOptions  swarm.SwarmOptions
+}
+
+func (provisioner *RedHatProvisioner) Service(name string, action pkgaction.ServiceAction) error {
+	command := fmt.Sprintf("sudo systemctl %s %s", action.String(), name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) Package(name string, action pkgaction.PackageAction) error {
+	var packageAction string
+
+	switch action {
+	case pkgaction.Install:
+		packageAction = "install"
+	case pkgaction.Remove:
+		packageAction = "remove"
+	case pkgaction.Upgrade:
+		packageAction = "upgrade"
+	}
+
+	switch name {
+	case "docker":
+		name = "docker"
+	}
+
+	command := fmt.Sprintf("sudo -E yum %s -y %s", packageAction, name)
+
+	if _, err := provisioner.SSHCommand(command); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (provisioner *RedHatProvisioner) isAWS() bool {
+	if _, err := provisioner.SSHCommand("curl -s http://169.254.169.254/latest/meta-data/ami-id"); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (provisioner *RedHatProvisioner) configureRepos() error {
+
+	// TODO: should this be handled differently? on aws we need to enable
+	// the extras repo different than a standalone rhel box
+
+	repoCmd := "subscription-manager repos --enable=rhel-7-server-extras-rpms"
+	if provisioner.isAWS() {
```

Comment:
Why not check whether we're on AWS by attempting a typecast of the driver to `amazonec2.Driver` like so:

```go
if _, err := provisioner.Driver.(amazonec2.Driver); err != nil {
    // do AWS-specific logic
} else {
    // do logic for everyone else
}
```

Admittedly this won't work for `generic`, so maybe we could fall back on the other method in that case.